src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs

```diff
@@ -0,0 +1,229 @@
+using System.Xml.Linq;
+using Microsoft.PowerFx.Core.App;
+using Microsoft.PowerFx.Core.App.ErrorContainers;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Binding.BindInfo;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Glue;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+using static Microsoft.PowerFx.Core.Localization.TexlStrings;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class UserDefinedFunction : TexlFunction
+    {
+        private readonly bool _isImperative;
+        private readonly IdentToken _returnTypeToken;
+        private readonly ISet<UDFArg> _args;
+
+        public TexlNode UdfBody { get; }
+
+        public override bool IsSelfContained => !_isImperative;
+
+        public UserDefinedFunction(string name, IdentToken returnType, TexlNode body, bool isImperative, ISet<UDFArg> args)
+        : base(DPath.Root, name, name, SG("Custom func " + name), FunctionCategories.UserDefined, returnType.GetFormulaType()._type, 0, args.Count, args.Count, args.Select(a => a.VarType.GetFormulaType()._type).ToArray())
+        {
+            this._returnTypeToken = returnType;
+            this._args = args;
+            this._isImperative = isImperative;
+        }
+
+        public TexlBinding Bind(INameResolver nameResolver, IBinderGlue documentBinderGlue, IUserDefinitionSemanticsHandler userDefinitionSemanticsHandler = null)
+        {
+            if (nameResolver is null)
+            {
+                throw new ArgumentNullException(nameof(nameResolver));
+            }
+
+            if (documentBinderGlue is null)
+            {
+                throw new ArgumentNullException(nameof(documentBinderGlue));
+            }
+
+            var parametersRecordType = CheckParameters(out var errors);
+            var bindingConfig = new BindingConfig(this._isImperative);
+            var binding = TexlBinding.Run(documentBinderGlue, UdfBody, UserDefinitionsNameResolver.Merge(nameResolver, parametersRecordType), bindingConfig);
+
+            CheckTypesOnDeclaration(binding.CheckTypesContext, _args, _returnTypeToken, binding.ResultType, binding.ErrorContainer);
+            userDefinitionSemanticsHandler?.CheckSemanticsOnDeclaration(binding, _args, binding.ErrorContainer);
+            binding.ErrorContainer.MergeErrors(errors);
+
+            return binding;
+        }
+
+        private RecordType CheckParameters(out List<TexlError> errors)
+        {
+            var record = RecordType.Empty();
+            var argsAlreadySeen = new HashSet<string>();
+            errors = new List<TexlError>();
+
+            foreach (var arg in _args)
+            {
+                if (!argsAlreadySeen.Add(arg.VarIdent.Name))
```

Comment:
This requires knowledge of how the Add function works and what the return value means. It would be better to call Contains on the hash set before trying to add anything.